File: segmentSupport/src/main/java/com/clumob/segment/support/appcompact/SegmentFragment.java
package com.clumob.segment.support.appcompact;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.clumob.segment.controller.SegmentController;
import com.clumob.segment.controller.SegmentInfo;
import com.clumob.segment.manager.Segment;
import com.clumob.segment.manager.SegmentManager;
import com.clumob.segment.manager.SegmentNavigation;
import com.clumob.segment.manager.SegmentViewHolder;

/**
 * Created by prashant.rathore on 14/02/18.
 */

public abstract class SegmentFragment extends Fragment implements SegmentManager.SegmentCallbacks {

    private Segment<?, ?> segment;
    private SegmentViewHolder viewHolder;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.segment = createSegment();
        segment.onCreate();
    }


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        segment.attach(getContext(), inflater);
        viewHolder = segment.createView(container);
        return viewHolder.getView();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        view.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View view, int i, KeyEvent keyEvent) {
                return segment.handleBackPressed();
            }
        });
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        segment.bindView(viewHolder);
    }

    @Override
    public void onStart() {
        super.onStart();
        segment.onStart();
    }

    @Override
    public void onResume() {
        segment.onResume();
        super.onResume();
    }

    @Override
    public void onPause() {
        segment.onPause();
        super.onPause();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onStop() {
        segment.onStop();
        super.onStop();
    }

    @Override
    public void onDestroyView() {
        this.getView().setOnKeyListener(null);
        this.segment.unBindView();
        this.viewHolder = null;
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        segment.onDestroy();
        super.onDestroy();
    }

    @Override
    public void onDetach() {
        this.segment = null;
        super.onDetach();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        segment.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        segment.onActivityResult(requestCode, resultCode, data);
    }

    protected Segment<?, ?> createSegment() {
        return provideSegment(provideSegmentInfo());
    }

    protected abstract SegmentInfo provideSegmentInfo();

    @Override
    final public void setSegmentView(View view) {
        throw new UnsupportedOperationException();
    }

    @Override
    final public SegmentNavigation createSegmentNavigation(SegmentManager segmentManager) {
        return null;
    }
}

File: segmentSupport/src/main/java/com/clumob/segment/support/appcompact/SegmentFragment.kt
package com.clumob.segment.support.appcompact

import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.clumob.segment.controller.SegmentInfo
import com.clumob.segment.manager.Segment
import com.clumob.segment.manager.SegmentManager
import com.clumob.segment.manager.SegmentManager.SegmentCallbacks
import com.clumob.segment.manager.SegmentNavigation
import com.clumob.segment.manager.SegmentViewHolder

/**
 * Created by prashant.rathore on 14/02/18.
 */
abstract class SegmentFragment : Fragment(), SegmentCallbacks {
    private var segment: Segment? = null
    private var viewHolder: SegmentViewHolder? = null
    override fun onAttach(context: Context) {
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        segment = createSegment()
        segment!!.onCreate()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        segment!!.attach(context!!, inflater)
        viewHolder = segment!!.createView(container)
        return viewHolder!!.view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.setOnKeyListener { view, i, keyEvent -> segment!!.handleBackPressed() }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        segment!!.bindView(viewHolder!!)
    }

    override fun onStart() {
        super.onStart()
        segment!!.onStart()
    }

    override fun onResume() {
        if (userVisibleHint) {
            segment!!.onResume()
        }
        super.onResume()
    }

    override fun onPause() {
        segment!!.onPause()
        super.onPause()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        segment!!.onConfigurationChanged(newConfig)
        super.onConfigurationChanged(newConfig)
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (view == null) {
            return
        }
        if (userVisibleHint) {
            segment!!.onResume()
        } else {
            segment!!.onPause()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
    }

    override fun onStop() {
        segment!!.onStop()
        super.onStop()
    }

    override fun onDestroyView() {
        this.view!!.setOnKeyListener(null)
        segment!!.unBindView()
        viewHolder = null
        super.onDestroyView()
    }

    override fun onDestroy() {
        segment!!.onDestroy()
        super.onDestroy()
    }

    override fun onDetach() {
        segment = null
        super.onDetach()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        segment!!.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        segment!!.onActivityResult(requestCode, resultCode, data)
    }

    protected fun createSegment(): Segment? {
        return provideSegment(provideSegmentInfo()!!)
    }

    protected abstract fun provideSegmentInfo(): SegmentInfo?
    override fun setSegmentView(view: View) {
        throw UnsupportedOperationException()
    }

    override fun createSegmentNavigation(segmentManager: SegmentManager): SegmentNavigation {
        return SegmentNavigation(segmentManager)
    }
}